### PR TITLE
Add timeout to MysqlDaemon.ReplicationStatus()

### PR DESCRIPTION
**Pull request by timvaillancourt**
## Description

This PR hopes to address https://github.com/vitessio/vitess/issues/11884 by adding a `context`-based timeout to `.ReplicationStatus()` method of `MysqlDaemon` _(in `go/vt/mysqlctl`)_

The lack of this timeout causes queries to be routed to unhealthy tablets in some situations

## Related Issue(s)

https://github.com/vitessio/vitess/issues/11884

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on the CI
-   [ ] Documentation was added or is not required

### CODE REVIEW

File: go/cmd/vtbackup/vtbackup.go [hunk 1]
```diff
@@ -362,7 +362,7 @@ func takeBackup(ctx context.Context, topoServer *topo.Server, backupStorage back
 		case <-time.After(time.Second):
 		}
 
-		status, statusErr := mysqld.ReplicationStatus()
+		status, statusErr := mysqld.ReplicationStatus(ctx)
```
Comment (brirams) [inline go/cmd/vtbackup/vtbackup.go:365]: cases like these are what scare me a bit in that we didn't have a timeout at all before and are now passing in whatever context was in scope. i think the only alternative would be to use `Context.TODO` in callsites that are unrelated to the our replication polling change but that sounds like a pain to keep track of, as i think it's directionally correct to have a context with a reasonable timeout throughout.

i think that's all to say: we should keep an eye on this other uses of the replication status check to ensure they continue to work as expected as we roll this out

**Comment by harshit-gangal** [comment]
do you plan to do this PR on OSS Vitess?

**Comment by timvaillancourt** [comment]
> do you plan to do this PR on OSS Vitess?

@harshit-gangal 👋 definitely once this PR is working. Right now these changes didn't help because it creates a loop on itself, just in a new place

Some old notes:
- I made `getPoolReconnect` use `executeFetchWithContext` (didn't use a `context` before) to run `SELECT 1` to test the conn
- `executeFetchWithContext` gets an error due to the `context`-timeout
- Then it runs `killConnection` to kill the connection
- `killConnection` uses `getPoolReconnect`, and we go back to step 1 🤦 

TL;DR: we get a stall, just in a new place. I plan to revisit this best-effort and appreciate any thoughts!

One thing that surprised me is `go/mysql` doesn't seem to use `context.Context`s (at least for queries) and functions like `executeFetchWithContext` is implementing a context timeout at a higher layer with a separate goroutine, etc. @harshit-gangal are there plans to add timeouts to `go/mysql`, or is this not-done intentionally?

**Comment by harshit-gangal** [comment]
> > > do you plan to do this PR on OSS Vitess?
> > 
> > 
> > @harshit-gangal 👋 definitely once this PR is working. Right now these changes didn't help because it creates a loop on itself, just in a new place
> > Some old notes:
> > 
> > * I made `getPoolReconnect` use `executeFetchWithContext` (didn't use a `context` before) to run `SELECT 1` to test the conn
> > * `executeFetchWithContext` gets an error due to the `context`-timeout
> > * Then it runs `killConnection` to kill the connection
> > * `killConnection` uses `getPoolReconnect`, and we go back to step 1 🤦
> > 
> > TL;DR: we get a stall, just in a new place. I plan to revisit this best-effort and appreciate any thoughts!
> > One thing that surprised me is `go/mysql` doesn't seem to use `context.Context`s (at least for queries) and functions like `executeFetchWithContext` is implementing a context timeout at a higher layer with a separate goroutine, etc. @harshit-gangal are there plans to add timeouts to `go/mysql`, or is this not-done intentionally?
> 
> The only way to kill query on mysql is to send `kill connection` command via separate connection. There is query killer that watches the connection timeout and kills those connection. We should access the underlying mysql using `dbconn` only that takes the context into consideration.

Added more thoughts here https://github.com/vitessio/vitess/issues/11884#issuecomment-1540292441
